bytestring/common/src/ByteString.kt

```diff
@@ -0,0 +1,456 @@
+     * wrapped by this byte string.
+     *
+     * @param startIndex the start index (inclusive) of a subsequence to copy, `0` by default.
+     * @param endIndex the end index (exclusive) of a subsequence to copy, [size] be default.
+     *
+     * @throws IndexOutOfBoundsException when [startIndex] or [endIndex] is out of range of byte string indices.
+     * @throws IllegalArgumentException when `startIndex > endIndex`.
+     */
+    public fun toByteArray(startIndex: Int = 0, endIndex: Int = size): ByteArray {
+        require(startIndex <= endIndex) { "startIndex: $startIndex, endIndex: $endIndex" }
+        return data.copyOfRange(startIndex, endIndex)
+    }
+
+    /**
+     * Copies a subsequence starting at [startIndex] and ending at [endIndex] of a byte sequence
+     * wrapped by this byte string and writes it into [destination] array starting at [destinationOffset] offset.
+     *
+     * @param destination the array to copy data into.
+     * @param destinationOffset the offset starting from which data copy should be written to [destination].
+     * @param startIndex the start index (inclusive) of a subsequence to copy, `0` by default.
+     * @param endIndex the end index (exclusive) of a subsequence to copy, [size] be default.
+     *
+     * @throws IndexOutOfBoundsException when [startIndex] or [endIndex] is out of range of byte string indices.
+     * @throws IndexOutOfBoundsException when the subrange doesn't fit into the [destination] array starting at
+     * the specified [destinationOffset], or when that index is out of the [destination] array indices range.
+     * @throws IllegalArgumentException when `startIndex > endIndex`.
+     */
+    public fun copyInto(
+        destination: ByteArray, destinationOffset: Int = 0,
+        startIndex: Int = 0, endIndex: Int = size
+    ) {
+        require(startIndex <= endIndex) {
+            "destinationOffset: $destinationOffset, startIndex: $startIndex, endIndex: $endIndex"
+        }
+        data.copyInto(destination, destinationOffset, startIndex, endIndex)
+    }
+
+    /**
+     * Returns a new byte string wrapping a subsequence of bytes wrapped by this byte string starting from
+     * [startIndex] and ending at [endIndex].
+     *
+     * @param startIndex the start index (inclusive) of a subsequence to copy.
+     * @param endIndex the end index (exclusive) of a subsequence to copy, [size] be default.
+     *
+     * @throws IndexOutOfBoundsException when [startIndex] or [endIndex] is out of range of byte string indices.
+     * @throws IllegalArgumentException when `startIndex > endIndex`.
+     */
+    public fun substring(startIndex: Int, endIndex: Int = size): ByteString = if (startIndex == endIndex) {
+        EMPTY
+    } else {
+        ByteString(data, startIndex, endIndex)
+    }
+
+    /**
+     * Compares a byte sequence wrapped by this byte string to a byte sequence wrapped by [other]
```

Comment:
It is worth adding either a more detailed description or examples of exactly how the comparison works.

If I am not mistaken, the lexicographic order does not strictly define the relation of values such as `(1, 2, 3) ? (1, 3)`

Reply:
I'll mention that the behavior is similar to `String::compareTo` and then add a sample comparing different byte strings in a separate PR (for https://github.com/Kotlin/kotlinx-io/issues/134)